_targets/build.project.ia32-generic

```diff
@@ -46,48 +90,69 @@ b_build_target() {
 	# Relink kernel to include newly created cpio archive
 	make -C "phoenix-rtos-kernel" all
 
-	# FIXME: build plo out of tree and without clean
-	b_log "Building plo"
-	make -C "plo/ia32" clean all
+	# Define default console device
+	if [ "$CONSOLE" = "serial" ]; then
+		PREINIT_SCRIPT+=("console 0.0")
+	else
+		PREINIT_SCRIPT+=("console 3.0")
+	fi
+
+	b_log "Building phoenix-rtos-loader"
+	b_mkscript_user USER_SCRIPT
+	b_mkscript_preinit
+	make -C plo base
 }
 
-b_build_project() {
-	b_log "Building user applications"
-	make -C "_user" all install
-}
 
 b_image_target() {
 	b_log "Creating image from $PREFIX_ROOTFS"
 
-	PLO="plo/ia32/plo"
-	KERNEL="$PREFIX_PROG/phoenix-ia32-generic.elf"
+	# Generate rootfs partition (ext2)
 	ROOTFS="$PREFIX_BUILD/disk.ext2"
-	IMG="$PREFIX_BUILD/phoenix-ia32-generic.disk"
-	IMG_DST="$PREFIX_BOOT/phoenix-ia32-generic.disk"
-
-	# partitions layout: (0kB)[MBR+plo] (32kB)[kernel] (2MB)[rootfs]
-	#FIXME: MBR should be constructed on per-project basic (here) instead of being embedded in plo
-
-	# first 2MBs: plo + kernel
-	dd of="$IMG" if=/dev/zero bs=2097152 count=1 >/dev/null 2>&1
-	dd of="$IMG" if="$PLO" bs=10486576 conv=notrunc >/dev/null 2>&1
-	dd of="$IMG" if="$KERNEL" seek=1 bs=32768 conv=notrunc >/dev/null 2>&1
-
-	# Adding ext2 rootfs
-	ext2_size=$((128 * 1024))
-	echo "Root filesystem size: ${ext2_size}KB"
-	genext2fs -b $ext2_size -i 2048 -d "$PREFIX_ROOTFS" "$ROOTFS"
-	dd of="$IMG" if="$ROOTFS" bs=2097152 seek=1 >/dev/null 2>&1
-
-	# produce image only if it's complete (no errors in the lines above)
-	mv "$IMG" "$IMG_DST"
+	rm -f "$ROOTFS"
+	SIZE_ROOTFS=$(find "$PREFIX_ROOTFS" -type f -exec du -k {} \; | awk 'BEGIN { sum=0 }; { sum+=$1; }; END { print sum }')
+	SIZE_ROOTFS=$((3 * SIZE_ROOTFS / 2))
+	echo "Root filesystem size: ${SIZE_ROOTFS}KB"
+	genext2fs -b $SIZE_ROOTFS -i 2048 -d "$PREFIX_ROOTFS" "$ROOTFS"
+
+	# Generate MBR partition table
+	PTABLE="$PREFIX_BUILD/disk.ptable"
+	rm -f "$PTABLE"
+	{
+		printf "\x80"
+		printf "\x0\x0\x0"
+		printf "\x83"
+		printf "\x0\x0\x0"
+		printf "\x$(printf "%x" $(((OFFS_ROOTFS >>  0) & 0xff)))"
+		printf "\x$(printf "%x" $(((OFFS_ROOTFS >>  8) & 0xff)))"
+		printf "\x$(printf "%x" $(((OFFS_ROOTFS >> 16) & 0xff)))"
+		printf "\x$(printf "%x" $(((OFFS_ROOTFS >> 24) & 0xff)))"
+		printf "\x$(printf "%x" $((((2 * SIZE_ROOTFS) >>  0) & 0xff)))"
+		printf "\x$(printf "%x" $((((2 * SIZE_ROOTFS) >>  8) & 0xff)))"
+		printf "\x$(printf "%x" $((((2 * SIZE_ROOTFS) >> 16) & 0xff)))"
+		printf "\x$(printf "%x" $((((2 * SIZE_ROOTFS) >> 24) & 0xff)))"
+	} > "$PTABLE"
```

Comment:
Maybe it would be easier to define a function like:
```
to_hex() {
	local num=$1 i
	for ((i=0; i<4; i++)); do
		printf "%s%02x" "\x" $(($num & 0xff))
		num=$(($num>>8))
	done
}
```
and then use:
```
	{
		printf $(to_hex $((0x80)))
		printf $(to_hex $((0x83)))
		printf $(to_hex $((OFFS_ROOTFS)))
		printf $(to_hex $((2 * SIZE_ROOTFS)))
	} > "$PTABLE"
```

Reply:
I agree, the code becomes more compact. I added the change in next force-push.